[14.0][IMP/ADD] stock_product_variant_mto, sale_stock_mto_as_mts_orderpoint_product_variant: Handle mto route at variant level

mmequignon opened the pull request:
I have to move commits in dedicated PRs, but pushed everything here for the time being

mmequignon committed:
sale_stock_mto_as_mts_orderpoint: Add hooks

mmequignon committed:
Add mto_route_product_variant

mmequignon committed:
Add sale_stock_mto_as_mts_orderpoint_product_variant

sebalix commented on mto_route_product_variant/__manifest__.py:
```diff
@@ -0,0 +1,18 @@
+
+{
+    "name": "MTO Route Product Variant",
+    "summary": "Allow to individually set variants as MTO",
```
```suggestion
    "summary": "Allow to individually set variants as Make To Order",
```

sebalix commented on mto_route_product_variant/models/product_product.py:
```diff
@@ -0,0 +1,63 @@
+from odoo import api, models, fields
+
+IS_MTO_HELP = """
+    Check or Uncheck this field if you want this variant to have another
+    mto configration than its template.
```
```suggestion
    Make To Order configration than its template.
```

grindtildeath replied:
The help is confusing.

To me, is_mto means the product has MTO route activated, whereas "another configuration than its template" does not mean that. I understand it as if the template is MTO, the variant is not, or if the template is not MTO, the variant is.

sebalix commented on mto_route_product_variant/models/product_product.py:
```diff
@@ -0,0 +1,63 @@
+    _inherit = "product.product"
+
+    is_mto = fields.Boolean(
+        string="Variant is Mto",
```
```suggestion
        string="Variant is MTO",
```

sebalix commented on mto_route_product_variant/__manifest__.py:
```diff
@@ -0,0 +1,18 @@
+# Copyright 2023 Camptocamp SA
```
Is `stock_product_variant_mto` a better name?

grindtildeath replied:
At least start with `stock_`.

grindtildeath commented on sale_stock_mto_as_mts_orderpoint_product_variant/__manifest__.py:
```diff
@@ -0,0 +1,16 @@
+{
```
Please add header

grindtildeath commented on mto_route_product_variant/models/product_product.py:
```diff
@@ -0,0 +1,63 @@
+        # mto disabled variants to be updated.
+        # To ensure that, the created `has_mto_route_changed` boolean field
+        # is set to True only when the MTO route is set on a template.
+        mto_route = self.env.ref("stock.route_warehouse0_mto")
```
```suggestion
        mto_route = self.env.ref("stock.route_warehouse0_mto", raise_if_not_found=False)
```

grindtildeath commented on mto_route_product_variant/models/product_product.py:
```diff
@@ -0,0 +1,63 @@
+        string="Variant is Mto",
+        compute="_compute_is_mto",
+        inverse="_inverse_is_mto",
+        store="true",
```
```suggestion
        store=True,
```

grindtildeath commented on mto_route_product_variant/models/product_product.py:
```diff
@@ -0,0 +1,63 @@
+        inverse="_inverse_is_mto",
+        store="true",
+        help=IS_MTO_HELP,
+        index=True
```
What's the reason for this? An index on a boolean column will almost never be used...

grindtildeath commented on mto_route_product_variant/models/product_template.py:
```diff
@@ -0,0 +1,24 @@
+        # This route_ids change will trigger variant's _compute_is_mto.
+        # We want to set variant's is_mto to True only if their
+        # template has been set to True here ↓
+        mto_route = self.env.ref("stock.route_warehouse0_mto")
```
```suggestion
        mto_route = self.env.ref("stock.route_warehouse0_mto", raise_if_not_found=False)
```

grindtildeath reviewed (commented):
I'm wondering if the issue at hand couldn't be solved by redefining `route_ids` as a computed field on `product.product` and then select the MTO route in the compute function only if the `is_mto` field is marked?

Maybe we should discuss this together.

grindtildeath committed:
Refactor mto_route_product_variant

Define route_ids as computed on the product variant, in order to
reuse route_ids from its template and add or remove MTO route
according to the setting on the variant.

In case MTO route is changed on the template, it must reset any
variant specific setting.

grindtildeath commented on stock_product_variant_mto/models/product_template.py:
```diff
@@ -1,24 +1,42 @@
+    @api.onchange("route_ids")
+    def onchange_route_ids(self):
+        mto_route = self.env.ref("stock.route_warehouse0_mto", raise_if_not_found=False)
+        if mto_route not in self._origin.route_ids and mto_route in self.route_ids._origin:
```
I guess we can make the condition more selective and raise the warning only if we have multiple variants, and these variants have different MTO settings.

grindtildeath commented on stock_product_variant_mto/models/product_template.py:
```diff
@@ -1,24 +1,42 @@
+                    "message": _("Activating MTO route will reset `Variant is MTO` setting on the variants.")
+                }
+            }
+        if mto_route in self._origin.route_ids and mto_route not in self.route_ids._origin:
```
Same as above

simahawk commented on stock_product_variant_mto/models/product_template.py:
```diff
@@ -1,24 +1,42 @@
+        templates_mto_after = self.filtered(lambda t: mto_route in t.route_ids)
+        templates_mto_added = template_not_mto_before & templates_mto_after
+        templates_mto_removed = (self - template_not_mto_before) & (self - templates_mto_after)
+        (templates_mto_added | templates_mto_removed).product_variant_ids._compute_is_mto()
```
`return res` missing

mmequignon committed:
fixup! Refactor mto_route_product_variant

simahawk commented:
@mmequignon to be reopened?

mmequignon commented:
> @mmequignon to be reopened?

yes please

jbaudoux commented on sale_stock_mto_as_mts_orderpoint_product_variant/models/product_product.py:
```diff
@@ -0,0 +1,41 @@
+        self.ensure_one()
+        return self.is_mto
+
+    def _inverse_is_mto(self):
```
Where is this defined on the field? I don't think this is called.
Check you have a unit test for this (archiving orderpoint when variant is not MTO anymore)

mmequignon commented:
I'll rollback @grindtildeath changes as it doesn't work.
Saving a backup with current code here https://github.com/camptocamp/stock-logistics-workflow/tree/handle_mto_route_at_variant_level_backup

mmequignon commented:
For reviewers, I know a test is failing, this is because in `sale_stock_mto_as_mts_orderpoint` mto products are enabled/disabled differently. I have to modify this, but feature works, tested from the ui.

grindtildeath committed:
[FIX] mto_as_mts_orderpoint_product_variant: Refactor for archival

There were multiple issues:

- Inverse function on storable field is not reliable and we should
prefer overriding write to add any processing when the field value
is changed. Moreover, the field is_mto on product.product did not
have its inverse function set on the field, so the function was not
called.

- When mto_as_mts is changed on the warehouse, we do not want to
archive MTO rule for this warehouse (because we could still want
to do MTO but not "as MTS").

henrybackman commented on sale_stock_mto_as_mts_orderpoint_product_variant/__manifest__.py:
```diff
@@ -0,0 +1,19 @@
+# Copyright 2024 Camptocamp SA
```
Inconsistent years 
```suggestion
# Copyright 2025 Camptocamp SA
```

henrybackman commented on sale_stock_mto_as_mts_orderpoint_product_variant/models/product_product.py:
```diff
@@ -0,0 +1,39 @@
+# Copyright 2023 Camptocamp SA
```
```suggestion
# Copyright 2025 Camptocamp SA
```

henrybackman commented on sale_stock_mto_as_mts_orderpoint_product_variant/models/product_product.py:
```diff
@@ -0,0 +1,39 @@
+        return res
+
+    def _get_orderpoints_to_archive_domain(self, warehouse):
+        # Orderpoints to archive are those where
```
Incomplete comment?

henrybackman commented on sale_stock_mto_as_mts_orderpoint_product_variant/tests/test_mto_as_mts_variant.py:
```diff
@@ -0,0 +1,155 @@
+# Copyright 2023 Camptocamp SA
```
```suggestion
# Copyright 2025 Camptocamp SA
```

henrybackman commented on stock_product_variant_mto/__manifest__.py:
```diff
@@ -0,0 +1,18 @@
+# Copyright 2023 Camptocamp SA
```
```suggestion
# Copyright 2025 Camptocamp SA
```

henrybackman commented on stock_product_variant_mto/models/product_product.py:
```diff
@@ -0,0 +1,52 @@
+# Copyright 2023 Camptocamp SA
```
```suggestion
# Copyright 2025 Camptocamp SA
```

henrybackman commented on stock_product_variant_mto/models/product_product.py:
```diff
@@ -0,0 +1,52 @@
+
+IS_MTO_HELP = """
+    Check or Uncheck this field to enable the Make To Order on the variant,
+    independantly from its template configuration.\n
```
```suggestion
    independently from its template configuration.\n
```

henrybackman commented on stock_product_variant_mto/models/product_template.py:
```diff
@@ -0,0 +1,44 @@
+# Copyright 2023 Camptocamp SA
```
```suggestion
# Copyright 2025 Camptocamp SA
```

henrybackman commented on stock_product_variant_mto/tests/common.py:
```diff
@@ -0,0 +1,80 @@
+# Copyright 2023 Camptocamp SA
```
```suggestion
# Copyright 2025 Camptocamp SA
```

henrybackman commented on stock_product_variant_mto/tests/test_mto_variant.py:
```diff
@@ -0,0 +1,108 @@
+# Copyright 2023 Camptocamp SA
```
```suggestion
# Copyright 2025 Camptocamp SA
```

henrybackman commented on stock_product_variant_mto/views/product_product.xml:
```diff
@@ -0,0 +1,32 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- Copyright 2023 Camptocamp SA
```
```suggestion
<!-- Copyright 2025 Camptocamp SA
```

henrybackman reviewed (approved):
Looks good :+1:

TDu commented on stock_product_variant_mto/models/product_product.py:
```diff
@@ -0,0 +1,52 @@
+        help=IS_MTO_HELP,
+    )
+
+    route_ids = fields.Many2many(
```
If find having this field on `product.product` for handling the feature of this module worrisome and could be conflicting with other modules or implementation...
